File: TripKitAndroid/src/main/java/com/skedgo/tripkit/DataLoader.java
package com.skedgo.tripkit;

import io.reactivex.Observable;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;
import io.reactivex.functions.Predicate;
import io.reactivex.schedulers.Schedulers;

import java.util.concurrent.Callable;
import java.util.concurrent.atomic.AtomicReference;


abstract class DataLoader<TData> implements Callable<Observable<TData>> {
  private final AtomicReference<TData> memoryCache = new AtomicReference<>();

  @Override
  public synchronized Observable<TData> call() {
    final Observable<TData> fromMemory = (memoryCache.get() == null) ? Observable.empty() : Observable.just(memoryCache.get());
    final Observable<TData> fromDisk = getDataAsync().subscribeOn(Schedulers.io());
    return Observable.concat(fromMemory, fromDisk)
            .filter(data -> data != null)
        .firstOrError().toObservable()
        .doOnNext(new Consumer<TData>() {
          @Override public void accept(TData data) {
            memoryCache.set(data);
          }
        });
  }

  public void invalidate() {
    memoryCache.set(null);
  }

  protected abstract Observable<TData> getDataAsync();
}

File: TripKitAndroid/src/main/java/com/skedgo/tripkit/DataLoader.kt
package com.skedgo.tripkit

import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers
import java.util.concurrent.Callable
import java.util.concurrent.atomic.AtomicReference

abstract class DataLoader<TData> : Callable<Observable<TData>> {
    val memoryCache = AtomicReference<TData>()

    @Synchronized
    override fun call(): Observable<TData> {
        val fromMemory = memoryCache.get().let { Observable.just(it) } ?: Observable.empty()
        val fromDisk = getDataAsync().subscribeOn(Schedulers.io())
        return Observable.concat(fromMemory, fromDisk)
            .filter { it != null }
            .firstOrError()
            .toObservable()
            .doOnNext { data -> memoryCache.set(data) }
    }

    fun invalidate() {
        memoryCache.set(null)
    }

    protected abstract fun getDataAsync(): Observable<TData>
}